src/java.base/share/native/libjli/java.c

```diff
@@ -534,6 +556,318 @@ invokeInstanceMainWithoutArgs(JNIEnv *env, jclass mainClass) {
+    int logFileLength = 0;
+    char *logFile = NULL;
+    logFilePropertyValue = getCommandLineOptionValue(LOG_FILE_OPTION_NAME, argc, argv, error);
+    if (!hasGetCommandLineOptionValueFoundError(*error)) {
+        if (NULL != logFilePropertyValue) {
+            logFileLength = snprintf(NULL, 0, "--log-file=%s", logFilePropertyValue) + 1;
+            if (logFileLength < 0) {
+                JLI_ReportErrorMessage("Failed to calculate the length of the CRIU log file string.");
+                freeCommandLineOptionValue(&logFilePropertyValue);
+                *error = MEMORY_ALLOCATION_ERROR;
+                return NULL;
+            }
+            logFile = (char *)JLI_MemAlloc(logFileLength);
+            if (NULL == logFile) {
+                JLI_ReportErrorMessage("Failed to allocate memory for the CRIU log file option.");
+                *error = MEMORY_ALLOCATION_ERROR;
+                return NULL;
+            }
+            snprintf(logFile, logFileLength, "--log-file=%s", logFilePropertyValue);
+            freeCommandLineOptionValue(&logFilePropertyValue);
+            return logFile;
+        } else {
+            JLI_ReportErrorMessage("The value of the command line option %s was not found.", LOG_FILE_OPTION_NAME);
+            *error = OPTION_VALUE_NOT_FOUND;
+            return NULL;
+        }
+    }
+    return NULL;
+}
+
+static int restoreFromCheckpoint(char *checkpointDir, char *logLevel, char *unprivilegedMode, char *logFile) {
+    int argc = 6;
+    char *argv[9] = { NULL };
+    argv[0] = "criu";
+    argv[1] = "restore";
+    argv[2] = "-D";
+    argv[3] = checkpointDir;
+    argv[4] = logLevel;
+    argv[5] = "--shell-job";
+    if (NULL != unprivilegedMode) {
+        argv[argc++] = unprivilegedMode;
+    }
+    if (NULL != logFile) {
+        argv[argc++] = logFile;
+    }
+    argv[argc] = NULL;
+    if (-1 == execvp(argv[0], argv)) {
+        return -1;
+    }
+    return 0;
+}
+
+static void handleCRaCRestore(int argc, char **argv) {
+    char *checkpointDir = NULL;
+    int error = 0;
+    char *unsupportedOption = NULL;
+    int parentProcessExitStatus = EXIT_SUCCESS;
+    pid_t childProcessPid = 0;
+    char *logLevel = NULL;
+    int childProcessExitStatus = EXIT_SUCCESS;
+    char *unprivilegedMode = NULL;
+    char *logFile = NULL;
+    int childProcessPidStatus = 0;
+    checkpointDir = getCheckpointDir(argc, argv, &error);
+    if (OPTION_NOT_FOUND == error) {
+        return;
+    }
+    if (hasGetCommandLineOptionValueFoundError(error)) {
+        JLI_ReportErrorMessage("Failed to get the CRIU checkpoint directory, error=%d.", error);
+        parentProcessExitStatus = EXIT_FAILURE;
+        goto exitParentProcessHandleCRaCRestore;
+    }
+    unsupportedOption = getUnsupportedCommandLineOption(argc, argv, &error);
+    if (OPTION_NOT_SUPPORTED == error) {
+        JLI_ReportErrorMessage("The command line option %s is not supported for CRaC restore.", unsupportedOption);
```

Comment:
This probably won't work if other non-CRIU but valid command line options exist.

Reply:
How can I determine if it is appropriate to remove the `OPTION_NOT_SUPPORTED` check given that restore occurs before jvm initialization?

Reply:
The options checked here are for `criu restore`, not for JVM, I think any unrecognized options can be ignored instead.